File: app/player_character.rb
class PlayerCharacter < GameObject
  RAMPS = [].freeze

  def initialize
    super(0, 0, 6, 6, :char, img_gap: Vector.new(-1, -1))
    @angle = 0
  end

  def set_position(col, row)
    @x = col * TILE_SIZE + 2
    @y = row * TILE_SIZE + 2
  end

  def update(scene)
    forces = Vector.new
    if KB.key_down?(:left_arrow)
      forces.x = -1
      @angle = -90
    elsif KB.key_down?(:right_arrow)
      forces.x = 1
      @angle = 90
    elsif KB.key_down?(:up_arrow)
      forces.y = -1
      @angle = 0
    elsif KB.key_down?(:down_arrow)
      forces.y = 1
      @angle = 180
    end
    move(forces, scene.obstacles_for(self), RAMPS, set_speed: true)
    scene.add_light(self, 3)
  end

  def draw
    super(angle: @angle, round: true)
  end
end

File: src/player_character.lua
PlayerCharacter = setmetatable({}, GameObject)
PlayerCharacter.__index = PlayerCharacter

function PlayerCharacter.new()
  local self = GameObject.new(0, 0, 6, 6, "char", Vector.new(-1, -1))
  setmetatable(self, PlayerCharacter)
  self.angle = 0
  return self
end

function PlayerCharacter:set_position(col, row)
  self.x = (col - 1) * TILE_SIZE + 2
  self.y = (row - 1) * TILE_SIZE + 2
end

function PlayerCharacter:update(scene)
  local forces = Vector.new()
  if KB.down("left") then
    forces.x = -1
    self.angle = -math.pi / 2
  elseif KB.down("right") then
    forces.x = 1
    self.angle = math.pi / 2
  elseif KB.down("up") then
    forces.y = -1
    self.angle = 0
  elseif KB.down("down") then
    forces.y = 1
    self.angle = math.pi
  end
  self:move(forces, scene:obstacles_for(self), scene.ramps, true)
  scene:add_light(self, 3)

  if self.on_exit then
    local exit_obj
    for _, e in ipairs(scene.exits) do
      if self:bounds():intersect(e) then
        exit_obj = e
        break
      end
    end
    if exit_obj then
      self.on_exit(exit_obj)
      return
    end
  end

  scene:check_triggers(self)
  scene:check_pushables(self)
end

function PlayerCharacter:draw()
  GameObject.draw(self, nil, nil, nil, self.angle, nil, nil, true)
end
